TypeCobol.LanguageServer/Workspace.cs

```diff
@@ -951,6 +952,183 @@ private void FinalCompilationStepCompleted(object cUnit, ProgramClassEvent progr
             DocumentModifiedEvent?.Invoke(fileUri, new EventArgs());
         }
 
+        #region Data Layout
+        private const char SPACE = ' ';
+        private const int INDENT_SIZE = 2; // Indentation = 2 spaces
+        private const string UNDEFINED = "***";
+        private const string FILLER = "FILLER";
+        private const string GROUP = "GROUP";
+        private const string OCCURS = "OCCURS {0}";
+        private const string PIC = "PIC {0}";
+        private const string REDEFINES = "REDEFINES {0}";
+
+        /// <summary>
+        /// Get the Data Layout rows for a Program or a Copy (output = CSV)
+        /// </summary>
+        /// <param name="compilationUnit">Compilation unit resulting from parsing the Program/Copy</param>
+        /// <param name="separator">Separator to be used in the rows</param>
+        /// <returns></returns>
+        public string[] GetDataLayoutAsCSV(CompilationUnit compilationUnit, string separator)
+        {
+            var rows = new List<string>();
+            foreach (var dataLayoutNode in CollectDataLayoutNodes(compilationUnit))
+            {
+                var row = CreateRow(dataLayoutNode, separator);
+                if (row != null)
+                {
+                    rows.Add(row);
+                }
+            }
+
+            return rows.ToArray();
+
+            static string CreateRow(Tuple<int, DataDefinition> dataLayoutNode, string separator)
+            {
+                var nodeLevel = dataLayoutNode.Item1;
+                var dataDefinition = dataLayoutNode.Item2;
+
+                var row = new StringBuilder();
+
+                //TODO manage slack bytes (property is dataDefinition.SlackBytes)
+
+                // Line number (starting at 1)
+                AppendToRow(dataDefinition.CodeElement.GetLineInMainSource() + 1);
+
+                /// Level number
+                AppendToRow(dataDefinition.CodeElement.LevelNumber);
+
+                // Name (preceded by an indent depending on the node level)
+                row.Append(SPACE, nodeLevel * INDENT_SIZE);
+                AppendToRow(GetName(dataDefinition));
+
+                // Declaration (Picture, Usage, REDEFINES, OCCURS, ...)
+                AppendDeclarationToRow(dataDefinition);
+
+                // Start/End/Length
+                var start = dataDefinition.StartPosition;
+                var length = dataDefinition.PhysicalLength;
+                AppendToRow(start);
+                AppendToRow(GetEnd(start, length));
+                AppendToRow(length);
+
+                return row.ToString();
+
+                void AppendToRow(object value)
+                {
+                    row.Append(value).Append(separator);
+                }
+
+                static string GetName(DataDefinition dataDefinition)
+                {
+                    return dataDefinition.Name == null ? FILLER : dataDefinition.Name;
```

Comment:
After test in real situation, dataDefinition under an occurs show the start and length of the first occurence of the array.
So the name should indicate that with a " (1)" suffix.
For example:
01.
    05 MyArray occurs 10.
      10 Var1 pic X.
       10 Var2 pic 9.
The result should be:

Line | Level | Name | Declaration | Start | End | Length
-- | -- | -- | -- | -- | -- | --
1 | 1 | FILLER | GROUP | 1 | 20 | 20
2 | 5 | MyArray | GROUP OCCURS 10 | 1 | 20 | 20
3 | 10 | Var1 (1) | PIC X | 1 | 1 | 1
4 | 10 | Var2 (1) | PIC 9 | 2 | 2 | 1

Reply:
A quick way to add the suffix.
[addSuffix.patch](https://github.com/TypeCobolTeam/TypeCobol/files/12427681/addSuffix.patch)

Reply:
Patch applied